### ISSUE TITLE
How To Fix An Illegal Turn, Non-stop Tight U-Turns, and Improve Routing? Almost got me into an accident..

### ISSUE THREAD

**Issue by grravity** [osmandapp/OsmAnd#10654]
For the first time today I tried to truly do a route to navigate using OSMAnd, and wow idk what happened but it did not go well.
Tried slightly 2 days ago, and it didn't go well either..

First off, just leaving the house, it took me out to the main road, and tried to make me do a super tight uturn which isn't possible in high traffic without going into the far lane.

After that I kept going, and all was smooth for about 5 minutes, until it tried to make me turn left where you legally cannot. I almost did before I noticing the sign, I had to go straight for another quarter mile then uturn where it'd dedicated to.

After I got to the parking lot of my destination which was a mall, I was driving through, and every two seconds it kept trying to redirect me while navigating the parking lot which was frustrating but whatever.

Now on the way back home it was terrible, it took me out, and immediately tried to make me do a uturn, I didn't because just like the last road, wayyy too tight, so I went the way I already knew which was to go straight and just take a left down the other road. If I didn't know the area I would have gotten completely lost. This is why I am testing in an area I know..
I kept going, and it constantly kept trying to take me on tight uturns that I can't make and screwing my whole route up.

It did this to me 2 days ago as well (tight uturn), and it almost got my into an accident. Luckily the guy saw me and hit the breaks.

I don't know if something is up with my settings, or I'm doing something wrong. I want to be able to use OSMAnd, I love it for everything else, but this experience has me slightly scarred especially as a newer driver.

For one, an option to avoid uturns where possible would be amazing, I don't know why it kept trying to do them, especially when it's not the ideal route. Incredibly annoying and dangerous.

Secondly, that illegal turn, I can't figure out how to submit a fix, or mark it so it doesn't do that.

Thirdly, the routing overall just seems very poor, based on what I said above, if there are suggestions please let me know.
I don't feel comfortable submitting a link here to the route publicly as it is my local area.

**Comment by scaidermern** [comment]
I think you need to add some examples / screenshots. This wall of text isn't very helpful.

**Comment by grravity** [comment]
Again, I don't really feel comfortable linking the routes.
I'll try to provide some screenshots when I get time.

But at least advice on how to fix the illegal left turn would be helpful.. I don't see why that'd need a screenshot.

**Comment by aceman444** [comment]
If there is an error in the data (a forbiddnen turn at some junction), you can fix that in the www.openstreetmap.org database from where OsmAnd takes the map data.

**Comment by scaidermern** [comment]
Right, you need to add a turn restriction. The default OSM editor iD has a helpful intersection visualization for this.

**Comment by grravity** [comment]
Here's example one of a super tight uturn on the main road, the first one it tried to do.
https://imgur.com/a/Wp3XGtX

The red arrows indicate the way it took me, the blue arrows indicate the safer way, although a bit longer. You take a left, then go straight and right around where the map cuts out there's a little rotary that takes you right back on that road going up, then you take that loop around that brings you to the other side of the road. 

Instead the map took me out to the right, and made me do a uturn. This is definitely shorter, however that uturn in person is super tight, and it's impossible to do it most of the time, as it is a main road with lots of traffic, and you'd be sitting there until both lanes are free because you can't make it into the inner lane.

Here's the second uturn it took me on on the way back. 
https://imgur.com/a/AR467T8

The route doesn't seem to show it, however this could have been because I was driving through the parking lot, and when I was doing so it said it was recalculating constantly. So I guess it recalculated into a route that made me uturn? I'm not very sure. I came up that road from the left, so I didn't go that route out of the mall. So I came up that main road, sat at that intersection, then it wanted me to uturn the way the arrows took me.
Again, this is a main road and pretty tight. This one almost got me into an accident. Of course my fault, but still. Why is it trying to constantly take me on uturns. There was about 3 other ways I could've gone.

I think maybe an option to avoid uturns would be nice. 
This route was only 2 miles, and I can't exactly imagine how many uturns it'd try to take me on on longer routes.
I can edit these to say nouturn on OSM, however I don't want to do that, because it's not actually illegal to uturn in those areas. Just not the safest/easiest idea most of the time.

Here's a screenshot of the illegal turn, 
https://imgur.com/a/zPw1lQk
So the proper way to fix is to click the road it took me down (the road at the top), and add a **relation** that says **no left turn; role: from**? Correct?

___________

_On a sidenote,_
How long does it take for OSMAnd to receive updates btw? I added some stores like 2 days ago to OSM, which are present while viewing OSM online, but they still haven't seemed to appear on OSMAnd?

**Comment by vshcherb** [comment]
So I guess it requires a thorough investigation. It's better always to share links like that to a location (https://www.openstreetmap.org/#map=19/50.30592/18.77062). If OSM has wrong navigation too you could share link like that https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=50.30530%2C18.76758%3B50.30609%2C18.76646#map=19/50.30577/18.76682. 

In OsmAnd it's possible to get Live updates (1 h delay) or create map yourself via OsmAndMapCreator.

**Comment by grravity** [comment]
Is there a place I can send a link of the route privately? An email? I understand it could be helpful, however again, didn't feel comfortable just posting that route publicly even with modified endpoints.

Also I thought I have live map updates enabled in settings, so unless it's a different setting?

**Comment by aceman444** [comment]
@grravity, at https://imgur.com/a/Wp3XGtX where the red arrows u-turn, is there a solid line on the road prohibiting the cars to u-turn or cross to the opposite lane?

**Comment by grravity** [comment]
@aceman444 https://imgur.com/a/9wcuB1n that's the u-turn you're asking about, so it looks like legally yeah it's prohibited. Didn't even go up that way because I knew I wouldn't be able to u-turn there, so I wasn't sure. So I should add a no u-turn on OSM?

Here's the other u-turn as well (the one I took that almost got me into the accident) https://imgur.com/vngzRIm

Neither of them have no u-turn signs, however the first one would be illegal, the second one is legal though correct? 
Either way, both are way too tight to safely take, and OSM routed me on both of them.

**Comment by aceman444** [comment]
@grravity thanks for the images. Yes in the first case as there is a solid line, a turn restriction should be added in OSM and OsmAnd (and no other app) can know if there is possible to turn or not. There are cases in the world where u-turn at such place is possible (see e.g. https://www.openstreetmap.org/way/147083442#map=19/40.61520/14.57528 connecting to Via Smeraldo). Existence of a sharp angle isn't a sufficient disqualification of a u-turn. Yes, if OsmAnd could guess that the connecting ways are different roads and not just a separately mapped directions (oneway) of the same road, that could be a solution but it is not clear to me how error-prone such guessing would be.

**Comment by grravity** [comment]
Marked that first one in OSM.

> Existence of a sharp angle isn't a sufficient disqualification of a u-turn.

Is there any way to mark this, and in turn in OSMAnd add an option to "avoid" sharp turns (if an alternative exists within a couple hundred-thousand feet)?
Or at least allow the user to mark it in OSMAnd, like just click on the part in the directions area ("Turn Left on XX Ave") then mark as "prefer to avoid" or look at alternative route options?

**Comment by aceman444** [comment]
You can try to set this in the routing.xml file in the OsmAnd data folder. There is a sharpTurn parameter for the "car" profile inside that file. Try to give it a large number (penalty) so that it avoids sharp turns.

**Comment by grravity** [comment]
I'll take a look when I get a chance, do you have a suggestion for what number?
Also could this be planned to be added to developer options maybe to easily adjust this in the app if this is something possible?

I am curious what it would consider a sharp uturn though, I'm thinking alternative options to pair with it such as avoiding uturns on main roads should be an option as well.

_______

Also I know this is unrealted, but does OSMAnd have the ability to select between routes? Most other mapping apps when selecting a destination, offer multiple routes, and they even show ones with toll or not for example. I know there is an option to avoid toll, but in general being able to select the better route when setting destination is ideal. Say a case where you know direction 2 is better than the default, or if you need to go over a toll and it is set to avoid by default the user can simply select the toll route otherwise it'd navigate them an extra 30min on the non toll route.

I can open a new request for this if wanted, but wanted to check if it already exists because I'm not seeing it, and it seems like a basic common feature so I might be missing it.

**Comment by aceman444** [comment]
No, OsmAnd does not suggest multiple alternative routes so far.

**Comment by Lee-Carre** [comment]
Re suggesting multiple routes: see #12103.

**Comment by hklene** [comment]
When reading this bug, I had to share my favorite Kamikaze U-Turn with you: Actually it is a U-turn, to do a 2nd U-turn, because it already anticipates you will not fully make the first U-turn on a 2-lane highway where people are speeding at 100 km/h both directions.

Lane advice explicitly tells me, not to go to the right - maybe the app was expecting the bridge to cave in?

![Screenshot_20180715-162932](https://user-images.githubusercontent.com/7427503/175348821-5f40754a-08eb-421c-b3a7-f139e7b88c56.jpg)

Here are the links:
https://www.openstreetmap.org/node/3703037421
https://www.google.de/maps/@52.7040315,8.2823614,265m/data=!3m1!1e3

It is already from 2018 ... I just keep it around, to tell people, not to trust software blindly :D

**Comment by vshcherb** [comment]
@hklene you can edit the map and put restriction and make the world better otherwise it will stay like this cause there is nothing wrong in software as it's wrong on the map which could  be supported by volunteers similar to you

**Comment by seniorm0ment** [comment]
Is there an easy way to mark these issues as you are driving?
Maybe that is a good idea. Because a lot of issues aren't apparent until you are actually driving, and it's not safe to try to fix it while driving.
So being able to mark it, then have a list of stuff you marked to fix would be great.
Like a "mark this area" or something.

**Comment by scaidermern** [comment]
> Is there an easy way to mark these issues as you are driving?

You can enable the OSM editing plugin and then create a map note at this location (tap on the map, choose actions / open OSM note).

However, it would be the best to log in with your OSM account first in the plugin settings. So that the note gets created under your account in case other mappers have further questions. And please always try to add as much information as possible. This note goes straight to OSM, nobody will know that you encountered a problem while using OsmAnd. So please always mention OsmAnd, where you came from and where you were going to, your mode of transportation (car, foot, bicycle).

So no, probably not an easy way, but at least there is a way :)

**Comment by aceman444** [comment]
> When reading this bug, I had to share my favorite Kamikaze U-Turn with you: Actually it is a U-turn, to do a 2nd U-turn, because it already anticipates you will not fully make the first U-turn on a 2-lane highway where people are speeding at 100 km/h both directions.
> 
> Lane advice explicitly tells me, not to go to the right - maybe the app was expecting the bridge to cave in?
> <img alt="" width="300" height="auto" src="https://user-images.githubusercontent.com/7427503/175348821-5f40754a-08eb-421c-b3a7-f139e7b88c56.jpg">
> 
> Here are the links: https://www.openstreetmap.org/node/3703037421 https://www.google.de/maps/@52.7040315,8.2823614,265m/data=!3m1!1e3
> 
> It is already from 2018 ... I just keep it around, to tell people, not to trust software blindly :D

This has been already fixed in the OSM data long ago.
